### PR TITLE
Containers are here !

### PR DESCRIPTION
**Pull request by FlorentinDUBOIS**

### CODE REVIEW

File: src/api.js [hunk 1]
```diff
@@ -43,6 +43,10 @@ class Api {
         .use(warp10)
         .use('/api', this.api);
 
+      if (process.env.PROXY_DOCKER_SOCKET) {
+        this.api.use('/container', require('./api/container'));
+      }
+
```
Comment (miton18) [inline src/api.js:49]: We have to decide, for Warp too, if we set env vars we use the middleware or if we always use the middleware and check in it the env var

File: src/api/container.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  .route('/')
+  .all(authenticationJwt)
+  .get(function(request, response) {
+    Docker
+      .list()
+      .then((dockers) => {
+        async.map(dockers, (docker, done) => {
+          docker
+            .information()
+            .catch(done)
```
Comment (miton18) [inline src/api/container.js:25]: Do nothing if container hasn't any informations ?

File: src/api/container.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      .catch((error) => {
+        response
+          .status(500)
+          .json({error: error.message});
+      });
+  })
+
+  .post(function(request, response) {
+    Docker
+      .run(request.body)
```
Comment (miton18) [inline src/api/container.js:51]: not necessary to check anything ?

File: src/api/container.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+          return response
+            .status(404)
+            .end();
+        }
+
+        return docker.information();
+      })
+
+      .then((information) => {
+        response.json({docker: information});
```
Comment (miton18) [inline src/api/container.js]: it's more "container" than "docker" itself ?

File: src/api/container.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        return docker.stop();
+      })
+
+      .then((docker) => {
+        return docker.remove();
+      })
+
+      .then((docker) => {
```
Comment (miton18) [inline src/api/container.js:113]: remove return container informations ? you can send it in response, like Routes

File: src/utils/docker/Docker.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      });
+    });
+  }
+
+  /**
+   * start the docker
+   * @return {Promise<Docker>} trigger when docker has stopped
+   */
+  start() {
+    return new Promise((resolve, reject) => {
```
Comment (miton18) [inline src/utils/docker/Docker.js:143]: When docker has stopped ?

**Comment by miton18** [comment]
src/utils/docker/DockerConnection.js is very necessary ?
Or with Singleton perhaps ?

**Review by miton18 — CHANGES_REQUESTED** [review]

File: src/api.js (hunk 1 above)
Comment (FlorentinDUBOIS) [reply src/api.js:49]: I prefer use like this, but we have to debate on it

File: src/api/container.js (hunk 2 above)
Comment (FlorentinDUBOIS) [reply src/api/container.js:25]: Container has always information the method information return a promise with the information given by the daemon

File: src/api/container.js (hunk 3 above)
Comment (FlorentinDUBOIS) [reply src/api/container.js:51]: Is no need because run has a validator to parameter see Docker.run method

File: src/api/container.js (hunk 4 above)
Comment (FlorentinDUBOIS) [reply src/api/container.js]: Right !

File: src/api/container.js (hunk 5 above)
Comment (FlorentinDUBOIS) [reply src/api/container.js:113]: Agreed

File: src/utils/docker/Docker.js (hunk 6 above)
Comment (FlorentinDUBOIS) [reply src/utils/docker/Docker.js:143]: Wrong copy/paste

**Comment by FlorentinDUBOIS** [comment]
DockerConnection is here to share the connection, like this the connection for me works like a singleton, it's not one but work like it.

File: src/api/container.js (hunk 2 above)
Comment (miton18) [reply src/api/container.js:25]: so why ctatch handler ?

File: src/api.js (hunk 1 above)
Comment (miton18) [reply src/api.js:49]: @TASK

File: src/main.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ const reporter    = require('./utils/reporter');
 const logger      = require('./utils/logger');
 const Joi         = require('joi');
 const {EnvSchema} = require('./models/env');
-idToPid           = new Map();
+const idToPid     = new Map();
```
Comment (miton18) [inline src/main.js:14]: will disappear

File: src/utils/docker/Docker.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    });
+  }
+
+  /**
+   * remove the docker
+   * @return {Promise<Docker>} trigger when docker has removed
+   */
+  remove() {
+    return new Promise((resolve, reject) => {
+      connection.getContainer(this.identifier).remove((error, information) => {
```
Comment (miton18) [inline src/utils/docker/Docker.js]: you can fix it ;-)

File: src/utils/docker/Image.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+  static pull(name) {
+    return new Promise((resolve, reject) => {
+      connection.pull(name, (error, stream) => {
+        if (error) {
+          logger.error(error.message);
+
+          return reject(error);
+        }
+
+        connection.modem.followProgress(stream, (error, outputs) => {
```
Comment (miton18) [inline src/utils/docker/Image.js]: you can fix it ;-)

**Review by miton18 — CHANGES_REQUESTED** [review]

File: src/api/container.js (hunk 2 above)
Comment (FlorentinDUBOIS) [reply src/api/container.js:25]: In case we can't access it

**Pull request merged**